proxyd/consensus_tracker.go

```diff
@@ -21,101 +26,311 @@ type ConsensusTracker interface {
+		ct.heartbeatInterval = heartbeatInterval
+	}
+}
+func NewRedisConsensusTracker(ctx context.Context,
+	redisClient *redis.Client,
+	bg *BackendGroup,
+	namespace string,
+	opts ...RedisConsensusTrackerOpt) ConsensusTracker {
+
+	tracker := &RedisConsensusTracker{
 		ctx:          ctx,
-		client:       r,
-		backendGroup: namespace,
+		client:       redisClient,
+		backendGroup: bg,
+		namespace:    namespace,
+
+		lockPeriod:        30 * time.Second,
+		heartbeatInterval: 2 * time.Second,
+		local:             NewInMemoryConsensusTracker().(*InMemoryConsensusTracker),
+		remote:            NewInMemoryConsensusTracker().(*InMemoryConsensusTracker),
+	}
+
+	for _, opt := range opts {
+		opt(tracker)
+	}
+
+	return tracker
+}
+
+func (ct *RedisConsensusTracker) Init() {
+	go func() {
+		for {
+			// follow same context as backend group poller
+			ctx := ct.backendGroup.Consensus.ctx
+			timer := time.NewTimer(ct.heartbeatInterval)
+			ct.stateHeartbeat()
+
+			select {
+			case <-timer.C:
+			case <-ctx.Done():
+				timer.Stop()
+				return
+			}
+		}
+	}()
+}
+
+func (ct *RedisConsensusTracker) stateHeartbeat() {
+	pool := goredis.NewPool(ct.client)
+	rs := redsync.New(pool)
+	key := ct.key("mutex")
+
+	val, err := ct.client.Get(ct.ctx, key).Result()
+	if err != nil && err != redis.Nil {
+		log.Error("failed to read the lock", "err", err)
+		if ct.leader {
+			ok, err := ct.redlock.Unlock()
+			if err != nil || !ok {
+				log.Error("failed to release the lock after error", "err", err)
+				return
+			}
+			ct.leader = false
+		}
+		return
+	}
+	if val != "" {
+		if ct.leader {
+			log.Debug("extending lock")
+			ok, err := ct.redlock.Extend()
+			if err != nil || !ok {
+				log.Error("failed to extend lock", "err", err, "mutex", ct.redlock.Name(), "val", ct.redlock.Value())
+				ok, err := ct.redlock.Unlock()
+				if err != nil || !ok {
+					log.Error("failed to release the lock after error", "err", err)
+					return
+				}
+				ct.leader = false
+				return
+			}
+			ct.postPayload(val)
+		} else {
+			// retrieve current leader
+			leaderName, err := ct.client.Get(ct.ctx, ct.key(fmt.Sprintf("leader:%s", val))).Result()
+			if err != nil && err != redis.Nil {
+				log.Error("failed to read the remote leader", "err", err)
+				return
+			}
+			ct.leaderName = leaderName
+			log.Debug("following", "val", val, "leader", leaderName)
+			// retrieve payload
+			val, err := ct.client.Get(ct.ctx, ct.key(fmt.Sprintf("state:%s", val))).Result()
+			if err != nil && err != redis.Nil {
+				log.Error("failed to read the remote state", "err", err)
+				return
+			}
+			if val == "" {
+				log.Error("remote state is missing (recent leader election maybe?)")
+				return
+			}
+			state := &ConsensusTrackerState{}
+			err = json.Unmarshal([]byte(val), state)
+			if err != nil {
+				log.Error("failed to unmarshal the remote state", "err", err)
+				return
+			}
+			ct.remote.mutex.Lock()
+			defer ct.remote.mutex.Unlock()
```

Comment:
Can we put the locking logic inside the remote state? Seems suboptimal to have to manage a struct's internal lock from outside the struct itself.

Reply:
Moved